### PR TITLE
Graph break on dynamic unspecialized ModuleList indexing

### LINKED ISSUE #122578: Recompiles and cache_size_limit from detectron2 CycleBatchNormList
Running
```
./benchmarks/dynamo/torchbench.py --inference --performance --only detectron2_fcos_r_50_fpn --stats --dynamic-shapes
```
I am seeing
```py
WARNING:torch._dynamo.convert_frame:torch._dynamo hit config.cache_size_limit (8)
   function: 'forward' (/home/jansel/conda/envs/pytorch/lib/python3.11/site-packages/detectron2/layers/batch_norm.py:318)
   last reason: L['self']._pos == 0      
```

The root cause is this class (slightly minified by hand):
``
class CycleBatchNormList(nn.ModuleList):
    def __init__(self, length: int, bn_class=nn.BatchNorm2d, **kwargs):
        super().__init__([bn_class(**kwargs, affine=False) for k in range(length)])
        self._pos = 0

    def forward(self, x):
        ret = self[self._pos](x)
        self._pos = (self._pos + 1) % len(self)
        return ret
```

TorchDynamo is correctly detecting the mutation of `self._pos` and mapping the class to `UnspecializedNNModuleVariable`, however the `self[self._pos]` is causing issues because it results in different weights being used each iteration.

This is a particularly tricky case to handle, since it is hard to prove that all values of `self._pos` will result in the same graph.  We also do not support indirect guards, so it is not possible to have a guard on `self[self._pos].blah`.

As a concrete action item, I think we should detect this type of pattern and graph break on it.  This would be a lot better than recompiling and hitting the cache limit, since it would allow us to optimize other parts of the model.

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186472

Fixes #122578

Dynamo realizes an unspecialized nn.Module attribute such as `_pos` as a guarded constant when it indexes a ModuleList-like container. If the attribute changes after the lookup, each value selects a different child and produces another guarded graph until the recompile limit is reached.

Track successful builtin nn.Module container lookups by their exact selector `Source` and inlined-frame ancestry for one analysis attempt. When the central side-effect path observes a matching non-idempotent mutation, restart analysis and raise a source-aware graph break at the lookup or deepest common caller. Fixed selectors remain compilable and exportable, and failed lookups are not treated as candidates.

The fallback deliberately uses recursive `SKIP/SKIP`. Allowing descendants to compile would let a called selector helper specialize independently before seeing the later mutation, recreating the original recompile loop.

Cache instance-specific skips with exact weak identities instead of applying them to the shared code object. Side-effect-free locators cover frame locals/globals and exact builtin list, tuple, dict, and instance-dict paths; the cache stores the actual strategy and reason, is weak in both code and instance keys, is shared by transient DDP backend clones, and is invalidated by reset and `reset_code`.

A broad graph break at every attribute-sourced lookup was rejected because it regresses fixed models and export. A code-global skip was rejected because it poisons fixed instances sharing forward code, while `SKIP/DEFAULT` was rejected because nested helpers recompile independently.

Test Plan:

- `python test/dynamo/test_modules.py -k module_list -v`
- `python test/dynamo/test_nested_graph_breaks_wrapped.py -k module_list -v`
- `python test/dynamo/test_autograd_function.py -k nullified_module_container_index -v`
- `python test/distributed/test_dynamo_distributed.py TestFakeDistributedSingleProc.test_ddp_optimizer_shares_instance_skip_cache -v`
- `python test/dynamo/test_modules.py NNModuleTests.test_nn_module_unspec_int_attr -v`
- `python test/dynamo/test_getitem.py -k nn_module -v`
- `python test/export/test_export.py TestExport.test_module_list_static_attr_index -v`
- `python test/export/test_export_strict.py -k module_list_static_attr_index -v`
- `python test/export/test_strict_export_v2.py -k module_list_static_attr_index -v`
- `lintrunner -a`
- `git diff --cached --check`

Generated by my agent

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says CI failed because the graph-break predicate also matched nn.utils.parametrize's closure key when indexing a ModuleDict. I narrowed the predicate to real unspecialized nn.Module attribute sources, excluded closure-cell sources, and added coverage for the intentional top-level frame skip while preserving parent-frame compilation around the dynamic container.

File: torch/_dynamo/variables/nn_module.py [hunk 1]
```diff
@@ -1095,6 +1096,50 @@ def unpack_var_sequence(
...
+    ) -> VariableTracker:
+        builtin_supported = (
+            torch.nn.ModuleDict.__getitem__,
+            torch.nn.ModuleList.__getitem__,
+            torch.nn.ParameterDict.__getitem__,
+            torch.nn.ParameterList.__getitem__,
+            torch.nn.Sequential.__getitem__,
+        )
+        getitem = getattr(type(self.value), "__getitem__", None)
+        if (
```
Comment (bobrenjc93) [inline torch/_dynamo/variables/nn_module.py:1284]: This gate has no mutation signal, so it fires far beyond the churn case the PR targets. `inline_inbuilt_nn_modules` is permanently True (config.py:573-577) and `mutation_guard.is_dynamic_nn_module` returns True for every non-export nn.Module (mutation_guard.py:121-124), and `wrap_module` explicitly routes `torch.nn.modules.container` types to `UnspecializedNNModuleVariable` (builder.py:2498-2511). That means any `self.container[<key read off module state>]` hits this branch, including keys that never change: `self.layers[self.idx]`, `self.heads[self.task_name]`, `self.blocks[self.cfg.layer_idx]`. Today those inline `__getitem__`, install a single EQUALS_MATCH guard on the attribute, and compile fullgraph; after this PR they graph break, and because `ConvertFrameReturn` defaults to `FrameExecStrategy(SKIP, DEFAULT)` with `apply_to_code=True`, the *code object* is skipped permanently — so every instance sharing that `forward` goes eager forever, even instances with a fixed index. Under `fullgraph=True`/`error_on_graph_break` the `Unsupported` is a hard error, i.e. a straight BC break for working code. Please gate on evidence that the key actually varies (e.g. `self.is_state_mutated`, a pending side-effect/STORE_ATTR on the key's source, or only when the key is already a dynamic `SymNodeVariable`) so the specialize-once path keeps working.

File: torch/_dynamo/variables/nn_module.py [hunk 2]
```diff
@@ -1095,6 +1096,50 @@ def unpack_var_sequence(
...
+            getitem in builtin_supported
+            and self._key_source_is_unspecialized_nn_module_attr(key)
+        ):
+            unimplemented(
```
Comment (bobrenjc93) [inline torch/_dynamo/variables/nn_module.py]: This also fires during export. `config.install_free_tensors_for_export` defaults to True (config.py:597) and is set on the export tracer (eval_frame.py:2452, functional_export.py:957), so `is_dynamic_nn_module(obj, is_export=True)` is True and container modules are unspecialized there too. Export cannot graph break or skip frames, so a model doing `self.layers[self.layer_idx](x)` will now fail export with this `Unsupported` instead of specializing as it does today. At minimum bail out early when `tx.output.export`, and add an export test covering container-indexed-by-module-attribute.

Comment (jansel) [reply torch/_dynamo/variables/nn_module.py]: My agent says fixed module-attribute indexing is exportable again. The patch no longer graph-breaks at lookup time; fixed cases pass torch._dynamo.export plus torch.export in non-strict, legacy strict, and strict-v2 modes. A genuinely mutating selector is rejected by legacy Dynamo export so it cannot silently specialize one child and lose the Python state mutation.

File: torch/_dynamo/variables/nn_module.py (hunk 1 above)
Comment (jansel) [reply torch/_dynamo/variables/nn_module.py:1284]: My agent says this is now gated on mutation evidence instead of every attribute-sourced lookup. Dynamo records the exact key Source and index ancestry, then only restarts when the central SideEffects store path observes a matching, non-deferred, value-changing selector write. Fixed and idempotent bool/int/str selectors still compile in fullgraph mode, and source-aware/identity-scoped fallback avoids contaminating fixed instances that share the same forward code. The new regression matrix covers direct/helper mutations, shared fixed+mutable forwards, idempotent stores, and repeated-call analysis counts.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased and updated this PR. The six actionable CI failures all came from the nested graph-break wrapper adding an extra expected first-pass frame; the tests now assert the intended invariant that analysis happens initially and does not recur. I also incorporated the review hardening: mutation-gated successful lookups, exact-weak instance caching with safe locators and reset invalidation, recursive helper skipping, DDP clone cache sharing, and focused ModuleDict, HOP, reset, lifetime, and aliasing coverage. The focused test matrix and lintrunner -a pass locally.

**Comment by jansel** [comment]
@claude review these changes

File: torch/_dynamo/variables/nn_module.py (hunk 1 above)
Comment (jansel) [reply torch/_dynamo/variables/nn_module.py:1284]: My agent says this is now gated on observed mutation rather than the lookup alone. Dynamo records only a successful builtin container lookup, then promotes it to a graph break only when the central side-effect path sees a matching non-idempotent store to the exact selector Source. Fixed instances still compile and export, and the negative skip cache is keyed by exact weak instance identity rather than the shared code object.

File: torch/_dynamo/variables/nn_module.py (hunk 2 above)
Comment (jansel) [reply torch/_dynamo/variables/nn_module.py]: My agent says fixed attribute indexing remains supported during export because the lookup alone no longer graph breaks. The new static-index compile/export regression tests pass; only a selector that is actually mutated after choosing the child is rejected, since exporting that case would otherwise specialize the wrong child and lose the Python state update.

**Comment by jansel** [comment]
My agent says
> The PR description now matches the current mutation-gated, instance-scoped implementation and expanded test plan; it no longer describes the superseded unconditional lookup-time graph break.
